realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java

```diff
@@ -307,6 +307,34 @@ private boolean categorizeClassElements() {
                     }
                 }
 
+                if (variableElement.getAnnotation(Required.class) == null) {
+                    // The field doesn't have the @Required annotation.
+                    // Without @Required annotation, boxed types/RealmObject/Date/String/bytes should be added to
+                    // nullableFields.
+                    // RealmList and Primitive types are NOT nullable always. @Required annotation is not supported.
+                    if (!Utils.isPrimitiveType(variableElement) && !Utils.isRealmList(variableElement)) {
+                        nullableFields.add(variableElement);
+                    }
+                } else {
+                    // The field has the @Required annotation
+                    if (Utils.isPrimitiveType(variableElement)) {
+                        Utils.error("@Required is not needed for field " + element +
+                                " with the type " + element.asType());
+                    } else if (Utils.isRealmList(variableElement)) {
+                        Utils.error("@Required is invalid for field " + element +
+                                " with the type " + element.asType());
+                    } else if (Utils.isRealmObject(variableElement)) {
+                        Utils.error("@Required is invalid for field " + element +
+                                " with the type " + element.asType());
+                    } else {
+                        // Should never get here - user should remove @Required
+                        if (nullableFields.contains(variableElement)) {
```

Comment:
It should never happen, at least, I don't expect it happen. But keep that error thrown here would give us some hint when it runs into this branch by some unknown reason. @cmelchior Maybe change the error message from "Consider to remove @Required" to "Please report this error to help@realm.io". Or just remove the "Consider to remove @Required." ?